File: Overlord.Core/Overlord.Core/Models/UpgradeModels.cs
using System.Text.Json.Serialization;

namespace Overlord.Core.Models;

/// <summary>
/// Weapon levels for Battle Cruisers.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum WeaponTier
{
    Laser,          // 1.0x damage (starting)
    Missile,        // 1.5x damage (Tier 2)
    PhotonTorpedo   // 2.0x damage (Tier 3)
}

/// <summary>
/// Upgrade costs and research times.
/// </summary>
public class UpgradeCosts
{
    /// <summary>
    /// Gets the research cost for a weapon tier.
    /// </summary>
    public static int GetResearchCost(WeaponTier tier)
    {
        return tier switch
        {
            WeaponTier.Missile => 500000,       // 500,000 Credits
            WeaponTier.PhotonTorpedo => 1000000, // 1,000,000 Credits
            _ => 0 // Laser is starting tier (no cost)
        };
    }

    /// <summary>
    /// Gets the research time in turns.
    /// </summary>
    public static int GetResearchTime(WeaponTier tier)
    {
        return tier switch
        {
            WeaponTier.Missile => 5,          // 5 turns
            WeaponTier.PhotonTorpedo => 8,    // 8 turns
            _ => 0 // Laser is starting tier (no research)
        };
    }

    /// <summary>
    /// Gets the damage modifier for a weapon tier.
    /// </summary>
    public static float GetDamageModifier(WeaponTier tier)
    {
        return tier switch
        {
            WeaponTier.Laser => 1.0f,
            WeaponTier.Missile => 1.5f,
            WeaponTier.PhotonTorpedo => 2.0f,
            _ => 1.0f
        };
    }
}

File: Overlord.Phaser/src/core/models/UpgradeModels.ts
import { WeaponTier } from './Enums';

/**
 * Upgrade costs and research times for weapon tiers.
 */
export class UpgradeCosts {
  /**
   * Gets the research cost for a weapon tier.
   */
  public static getResearchCost(tier: WeaponTier): number {
    switch (tier) {
      case WeaponTier.Missile:
        return 500000; // 500,000 Credits
      case WeaponTier.PhotonTorpedo:
        return 1000000; // 1,000,000 Credits
      default:
        return 0; // Laser is starting tier (no cost)
    }
  }

  /**
   * Gets the research time in turns.
   */
  public static getResearchTime(tier: WeaponTier): number {
    switch (tier) {
      case WeaponTier.Missile:
        return 5; // 5 turns
      case WeaponTier.PhotonTorpedo:
        return 8; // 8 turns
      default:
        return 0; // Laser is starting tier (no research)
    }
  }

  /**
   * Gets the damage modifier for a weapon tier.
   */
  public static getDamageModifier(tier: WeaponTier): number {
    switch (tier) {
      case WeaponTier.Laser:
        return 1.0;
      case WeaponTier.Missile:
        return 1.5;
      case WeaponTier.PhotonTorpedo:
        return 2.0;
      default:
        return 1.0;
    }
  }
}
